File: cses/09_geometry/2190_line-segment-intersection/solution.py
# Line Segment Intersection - https://cses.fi/problemset/task/2190
# Xác định hai đoạn thẳng có giao nhau hay không (có chung ít nhất một điểm).
# Dùng cross product (orientation) với số nguyên chính xác tuyệt đối.

import sys


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    t = data[0]
    i = 1
    out = []
    ap = out.append

    for _ in range(t):
        x1 = data[i]; y1 = data[i + 1]
        x2 = data[i + 2]; y2 = data[i + 3]
        x3 = data[i + 4]; y3 = data[i + 5]
        x4 = data[i + 6]; y4 = data[i + 7]
        i += 8

        # Vector chỉ phương của mỗi đoạn (tính trước để tái sử dụng).
        ax = x4 - x3; ay = y4 - y3   # p3 -> p4
        bx = x2 - x1; by = y2 - y1   # p1 -> p2

        # cross((p2-p1) x (q-p1)) — orientation of q relative to a directed line.
        # d1, d2: vị trí của p1, p2 so với đường thẳng p3p4.
        # d3, d4: vị trí của p3, p4 so với đường thẳng p1p2.
        d1 = ax * (y1 - y3) - ay * (x1 - x3)  # p1 vs line p3p4
        d2 = ax * (y2 - y3) - ay * (x2 - x3)  # p2 vs line p3p4
        d3 = bx * (y3 - y1) - by * (x3 - x1)  # p3 vs line p1p2
        d4 = bx * (y4 - y1) - by * (x4 - x1)  # p4 vs line p1p2

        # Trường hợp giao nhau "thật sự" (proper): p1, p2 nằm khác phía đối với
        # đường thẳng p3p4 VÀ p3, p4 nằm khác phía đối với đường thẳng p1p2.
        if ((d1 > 0) != (d2 > 0)) and d1 != 0 and d2 != 0 and \
           ((d3 > 0) != (d4 > 0)) and d3 != 0 and d4 != 0:
            ap("YES")
            continue

        # Trường hợp collinear / chạm đầu mút: một điểm thẳng hàng (d == 0)
        # và nằm trong bounding box của đoạn kia (kiểm tra on-segment).
        if d1 == 0 and min(x3, x4) <= x1 <= max(x3, x4) and min(y3, y4) <= y1 <= max(y3, y4):
            ap("YES")
        elif d2 == 0 and min(x3, x4) <= x2 <= max(x3, x4) and min(y3, y4) <= y2 <= max(y3, y4):
            ap("YES")
        elif d3 == 0 and min(x1, x2) <= x3 <= max(x1, x2) and min(y1, y2) <= y3 <= max(y1, y2):
            ap("YES")
        elif d4 == 0 and min(x1, x2) <= x4 <= max(x1, x2) and min(y1, y2) <= y4 <= max(y1, y2):
            ap("YES")
        else:
            ap("NO")

    sys.stdout.write("\n".join(out) + "\n")


main()

File: cses/09_geometry/2190_line-segment-intersection/solution.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>

using i64 = std::int64_t;
using i128 = __int128_t;

struct Point {
    i64 x;
    i64 y;
};

// Cross product (b - a) x (c - a): dấu cho biết orientation của điểm c so với
// hướng a -> b (trái > 0, phải < 0, thẳng hàng = 0). Dùng __int128 tránh tràn số.
[[nodiscard]] i128 cross(const Point& first, const Point& second, const Point& third) {
    return static_cast<i128>(second.x - first.x) * (third.y - first.y) -
           static_cast<i128>(second.y - first.y) * (third.x - first.x);
}

// Kiểm tra on-segment bằng bounding box: điểm (đã thẳng hàng) có nằm trong hình
// chữ nhật bao của đoạn first-second theo cả hai trục x và y hay không.
[[nodiscard]] bool on_segment(const Point& first, const Point& second, const Point& point) {
    return std::min(first.x, second.x) <= point.x && point.x <= std::max(first.x, second.x) &&
           std::min(first.y, second.y) <= point.y && point.y <= std::max(first.y, second.y);
}

// Hai giá trị trái dấu (đều khác 0): dùng để xét hai đầu mút nằm khác phía.
[[nodiscard]] bool opposite_signs(i128 first, i128 second) {
    return (first < 0 && second > 0) || (first > 0 && second < 0);
}

// Xác định hai đoạn first-second và third-fourth có giao nhau hay không.
[[nodiscard]] bool intersect(const Point& first, const Point& second,
                             const Point& third, const Point& fourth) {
    // Bốn orientation: vị trí của third, fourth so với đường thẳng first-second
    // và vị trí của first, second so với đường thẳng third-fourth.
    const i128 third_side = cross(first, second, third);
    const i128 fourth_side = cross(first, second, fourth);
    const i128 first_side = cross(third, fourth, first);
    const i128 second_side = cross(third, fourth, second);

    // Giao nhau "thật sự" (proper): mỗi đoạn có hai đầu mút nằm khác phía so với
    // đường thẳng chứa đoạn kia -> cắt nhau tại đúng một điểm bên trong.
    if (opposite_signs(third_side, fourth_side) && opposite_signs(first_side, second_side)) {
        return true;
    }
    // Trường hợp biên (collinear / chạm đầu mút): một điểm thẳng hàng với đoạn kia
    // (orientation = 0) và nằm trong bounding box của đoạn đó.
    if (third_side == 0 && on_segment(first, second, third)) {
        return true;
    }
    if (fourth_side == 0 && on_segment(first, second, fourth)) {
        return true;
    }
    if (first_side == 0 && on_segment(third, fourth, first)) {
        return true;
    }
    return second_side == 0 && on_segment(third, fourth, second);
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    // Đọc số truy vấn rồi xử lý từng truy vấn trong O(1).
    int test_count = 0;
    std::cin >> test_count;
    while (test_count-- > 0) {
        Point first{};
        Point second{};
        Point third{};
        Point fourth{};
        std::cin >> first.x >> first.y >> second.x >> second.y >>
            third.x >> third.y >> fourth.x >> fourth.y;
        std::cout << (intersect(first, second, third, fourth) ? "YES\n" : "NO\n");
    }
}
